CRM/Birthdays/Mailer.php

```diff
@@ -0,0 +1,121 @@
+<?php
+/*
+ * Copyright (C) 2022 SYSTOPIA GmbH
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as published by
+ *  the Free Software Foundation in version 3.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use Civi\API\Exception\UnauthorizedException;
+use Civi\Api4\Activity;
+use CRM_Birthdays_ExtensionUtil as E;;
+
+class CRM_Birthdays_Mailer
+{
+    private int $template_id;
+    private string $email_address_from;
+
+    /**
+     * @throws Exception
+     */
+    public function __construct()
+    {
+        $this->template_id = Civi::settings()->get(CRM_Birthdays_Form_Settings::BIRTHDAYS_MESSAGE_TEMPLATE);
+        if (empty($this->template_id)) {
+            throw new Exception('Birthdays: Message template not found. Please set a template in birthday settings');
+        }
+
+        $email_id = Civi::settings()->get(CRM_Birthdays_Form_Settings::BIRTHDAYS_SENDER_EMAIL_ADDRESS_ID);
+        $this->email_address_from = $this->get_sender_email_address_from_id($email_id);
+        if (empty($this->email_address_from)) {
+            throw new Exception('Birthdays: Pre selected outgoing email not found. Please set am outgoing email address in birthday settings');
+        }
+    }
+
+    /**
+     * @param $contacts
+     * @param $write_activity
+     * @return int error count
+     * @throws API_Exception
+     * @throws CRM_Core_Exception
+     * @throws UnauthorizedException
+     */
+    public function send_mails_and_write_activity($contacts, $write_activity): int
+    {
+        $error_count = 0;
+        foreach ($contacts as $contact_id => $contact_info) {
+            try {
+                $this->send_mail($contact_id, $this->email_address_from, $contact_info['email'], $this->template_id);
+                if ($write_activity) $this->create_activity($contact_id, E::ts('Successful birthday greeting mail'), E::ts('Successful birthday greeting mail! Template ID %1 has been used.', [$this->template_id]));
```

Comment:
`ts()` parameters must be indexed with their corresponding `%#` number, i.e. `%1` will need an array key `1` for `$this->template_id`, and parameters are `1`-based, as opposed to `0`-based implicit array indices.